### PR TITLE
Jonmv/respect exclusivity in hcm

### PR DESCRIPTION
**Pull request by jonmv**
@freva and/or @hakonhall please review

### CODE REVIEW

**Review by freva — APPROVED** [review]

File: node-repository/src/main/java/com/yahoo/vespa/hosted/provision/provisioning/NodeCandidate.java [hunk 1]
```diff
@@ -559,4 +563,32 @@ public String toString() {
 
     }
 
+    public boolean violatesExclusivity(ClusterSpec cluster, ApplicationId application,
+                                       boolean exclusiveCluster, boolean hostSharing, NodeList allNodes) {
+        if (parentHostname().isEmpty()) return false;
+        if (type() != NodeType.tenant) return false;
+
+        // We always violate exclusivity if the parent is exclusive to someone else that the requesting application.
+        if ( ! emptyOrEqual(parent.flatMap(Node::exclusiveToApplicationId), application)) return true;
```
Comment (hakonhall) [inline node-repository/src/main/java/com/yahoo/vespa/hosted/provision/provisioning/NodeCandidate.java:572]: Why was this moved out from within the !hostSharing condition ?

Comment (jonmv) [reply node-repository/src/main/java/com/yahoo/vespa/hosted/provision/provisioning/NodeCandidate.java:572]: This was to ensure hosts with `exclusiveTo` set would _never_ be reused by someone else (including in zones where host-sharing is not allowed, but where the host is empty).

Comment (hakonhall) [reply node-repository/src/main/java/com/yahoo/vespa/hosted/provision/provisioning/NodeCandidate.java:572]: Moving it out of the !hostSharing condition doesn't affect zones where host-sharing is not allowed.  Host-sharing here means YAHOO.

This is likely a no-op because hosts do not have exclusiveToApplicationId in YAHOO.  Same with exclusiveToClusterType.

My point is that the 3 conditions under !hostSharing belong together, so either we should keep them under !hostSharing, or move all out.

Comment (jonmv) [reply node-repository/src/main/java/com/yahoo/vespa/hosted/provision/provisioning/NodeCandidate.java:572]: Then I think we should move them all out. It had me confused, at least.
